### ISSUE TITLE
Some Enums are not Renamed Without (Obvious) Reason

### ISSUE THREAD

**Issue by ab-tools** [yck1509/ConfuserEx#364]
Sorry to bother you with different issues at once, but I'm currently in the progress to completely switch to ConfuserEx for a bigger project. ;-)

I've no idea why, but as you can see in this screenshot there are some (precisely 5) enums that are not renamed and I have no idea why:
![confuserex-someenumsnotrenamed](https://cloud.githubusercontent.com/assets/1347603/11162945/3aa5525a-8abb-11e5-825c-19fd50fb1c3b.png)

I really have a lot enums across the whole project and they are all renamed perfectly fine (including namespace flattened to "root") except these 5 and I don't find a difference.

I did not add any obfuscation exceptions till now, just added this single line before my top namespace "namespace FlightControl" - nothing else:
[assembly: Obfuscation(Exclude = false, Feature = "anti debug;anti ildasm;constants(elements=SNPI);rename(mode=sequential)")]

Just checked and all 5 enums that are not renamed in the namespace "FlightControl.FsImport" are defined as "internal enum" - so there also should not be an issue regarding external visibility for public enums.

Any idea what the reason could be that these enums are not renamed?

Thank you again very much in advance for great support
Andreas

**Comment by yck1509** [comment]
Hi,

Most likely you had unknowingly used Reflection with these enums. You could try checking your project for typeof, Enum.ToString/Parse, etc. that reference the enum.

**Comment by ab-tools** [comment]
I'm a bit confused as even if I would use refelction on these enums (which I don't think I do, but will check again), it should not be a problem to rename the enum type name itself, is it?

Only the enum elements must not be renamed when reflection is used.
Or do I missunderstand something here?

Thank you
Andreas

**Comment by yck1509** [comment]
For example, you might serialize a enum, which the name of enum is required to deserialize it. You might call typeof(Enum).FullName to get its name, or call Type.GetType("EnumName"). ConfuserEx already handled many parts of Reflection quite well, but it might have false positive and didn't rename your enum even though it could.

**Comment by ab-tools** [comment]
Then this really seems a "false positive" as this example of "AirspaceType" shows:
![confuserex-enum](https://cloud.githubusercontent.com/assets/1347603/11169631/037a927e-8bbd-11e5-8bde-6d3502a51202.png)
As you can see these are all 26 references project-wide found by Visual Studio. None involves reflection.

But I guess I can work-around it with the "forceRen" rename option.

**Comment by yck1509** [comment]
Hmm... I suppose this may be a false positive. If you are able to make a simple sample that shows the issue, I could try to fix it.

**Comment by ab-tools** [comment]
OK, I'll try tomorrow, but I'm not sure if that's possible:
I see no reason why only these enums are not renamed as there are a lot others which are. I would need to find that out first to "build" a simple sample showing the problem...

**Comment by ab-tools** [comment]
I'm still trying to figure out what the problem is here, but I was not able till now to build a "small project" where this issue happens, too.

First I've removed the two lines of code you suggested - just to be sure:
https://github.com/yck1509/ConfuserEx/issues/363

But the strangest issue is that Confuser refuses to obfuscate these 5 enums above even if I directly place a
    [Obfuscation(Exclude = false, Feature = "rename(mode=sequential,forceRen=true)")]
in front of them! It just ignores them whatever I do...

For the assembly itself I just use this obfuscation attribute:
    [assembly: Obfuscation(Exclude = false, Feature = "anti debug;anti ildasm;constants(elements=SNPI);rename(mode=sequential)")]

Do you have any idea what the problem might be?
Or is there a way to enable more logging for Confuser so that it outputs why it includes/excludes things from obfuscation?

Thanks again
Andreas

**Comment by ab-tools** [comment]
Just wanted to let you know that I found the reason now why these specifc (few) enums were not renamed:
Deeply hidden in the big source code of the project there were really a few times ".ToString()" used - only on this specific enums. Therefore ConfuserEx completely understandable decided to better not rename these enums. :-)

I fixed that now (removed all ".ToString()" calls on these enums) and suddenly these enums are renamed, too!

This way I can close this issue now.
Just wanted to let you know about the reason first. Thanks again!
